Docs / debugging (#4578)
canonical/cloud-init: doc/rtd/explanation/failure_states.rst

.. _failure_states:

Failure states
==============

> 

Author: > It's not ideal to have sequential headers with no text in between them

Agreed, fixed.

> Could we take a second look at the header structure to avoid that?

I've broken some of this content out into two new pages:
```
doc/rtd/explanation/exported_errors.rst
doc/rtd/howto/status.rst
```

Critical failure
----------------
If cloud-init is unable to complete, error messages will now

> additionally be visible in output of `cloud-init status --long`

Author: I think actually that this wording should be removed rather than enumerating other less-user-friendly interfaces

> additionally be visible in output of `cloud-init status --long`

Reviewer: That would also work! :)

within the 'error' key nested under the module-level keys: 'init-local',
'init', 'modules-config', 'modules-final'.

Recoverable failure
-------------------
In the case that cloud-init is able to complete yet something goes awry,

> the service will now return with exit code 2, and error messages will be

Author: Documenting former state and current state would reduce the value of this documentation, I think, and would maybe have made more sense if former state was actually documented prior (it wasn't). I'll remove this wording.

visible in the output of `cloud-init status --long` under the top
level 'recoverable_errors' key as well as within the 'error' key nested
under the module-level keys: 'init-local', 'init', 'modules-config',

> 'modules-final'.

Author: Let me know if the rewording is better

> 'modules-final'.

Reviewer: Much clearer, thank you!

Implementation
==============

Cloud-init error codes
----------------------

> 0 - success

Author: > I get the impression this might also need some introductory text - without context I'm a bit confused about whether these are universal error codes that are applied to all failures, or if they're more specific than that and only apply in certain cases.

+1, will do

> We may also want to indent these as a quote block to avoid them all being shown on the same line. My preference would be to have them in a bulleted list, with the code (and meaning) on the bulleted line in monospaced font, then the corresponding text under that bullet point on the next line (to make it clear where it belongs).

I think this comment applies to a prior version of this change. Could you please let me know if the updated version looks better?

1 - unrecoverable error
2 - recoverable error

If cloud-init exits with exit code 1, cloud-init experienced critical failure
and was unable to recover. In this case, something is likely seriously
wrong with the system, or cloud-init has experienced a serious bug.

If cloud-init exits with exit code 2, cloud-init was able to complete

> gracefully, however something went wrong and the user should investigate.

Author: "complete gracefully" means cloud-init was able to "finish, but not do what expected"

> gracefully, however something went wrong and the user should investigate.

Author: Lots more improvements on this page and the two new ones - hopefully this is now more clearly defined?

Reported state
--------------
Cloud-init can report its internal state via the `status --format json`
subcommand under the `'extended_status'` key.

.. code-block: shell-session

    $ cloud-init status --format json | jq .extended_status
    "degraded done"

See the list of all possible states:

.. code-block: shell-session

  "not running"
  "running"
  "done"
  "error"
  "degraded done"
  "degraded running"
  "disabled"

Exported errors: Aggregated errors
----------------------------------

> When a recoverable error occurs, the internal cloud-init state

Author: > We haven't defined anywhere what a "recoverable error" is 

See the "error codes" section

> and it doesn't match the section heading, either.

This section only applies to recoverable errors, not unrecoverable errors, should that be clarified maybe?

> When a recoverable error occurs, the internal cloud-init state

Reviewer: Looking back to where you pointed, I think the definition of a recoverable error is fine (although thinking about it, I'd maybe put subsubheadings for the recoverable/non-recoverable error definitions up there to call those sections out specifically). The primary confusion for me is how recoverable/non-recoverable errors relate to Exported and Aggregated errors (as in the headings) since this isn't immediately obvious.

> When a recoverable error occurs, the internal cloud-init state

Author: The exported and aggregated errors content relates to recoverable errors. I've broken this content out into a separate page to make that more clear.

information is made visible under a top level aggregate key
'recoverable_errors' with errors sorted by error level:

.. code-block: shell-session

    $ cloud-init status --format json | jq .recoverable_errors
    {
      "WARNING": [
        "Failed at merging in cloud config part from p-01: empty cloud config",
        "No template found in /etc/cloud/templates for template source.deb822",
        "No template found in /etc/cloud/templates for template sources.list",
        "No template found, not rendering /etc/apt/soures.list.d/ubuntu.source"
      ]
    }

Reported recoverable error messages are grouped by the level at which
they are logged. Complete list of levels:

.. code-block: shell-session

  WARNING
  DEPRECATED
  ERROR

>   CRITICAL

Author: I think I've answered the other ones, but not sure how to word an answer for this one:

> Do all warnings get reported, or can they be suppressed if a more serious one overlaps it?

All warnings are reported, but there is no "overlap".

>   CRITICAL

Author: More detail added (now in exported_errors.rst)

Exported errors: Per-stage errors
---------------------------------
The keys 'errors' and 'recoverable_errors' are also exported for each
stage to allow attribution of recoverable and non-recoverable errors

> to their source.

Author: see new wording

> to their source.

Reviewer: Much more straightforward, thank you!

.. code-block: shell-session

    $ cloud-init status --format json | jq .init.recoverable_errors
    {
      "WARNING": [
        "Failed at merging in cloud config part from p-001: empty cloud config"
      ]
    }

Note: Only cloud-init stages which have completed are listed in the
output of `cloud-init status --format json`.

The json representation of cloud-init stages (in run order) is:

.. code-block: shell-session

  "init-local"
  "init"
  "modules-config"
  "modules-final"

Limitations of internal errors
==============================
- Exported recoverable errors represent logged messages, which are not
  guaranteed to be stable between releases. The contents of the
  'errors' and 'recoverable_errors' keys are not guaranteed to have
  stable output!
- Exported errors and recoverable errors may occur at different stages
  since users may reorder configuration modules to run at different

>   stages via cloud.cfg.

Author: fixed, I think

the service will now return with exit code 2, and error messages will be
visible in the output of `cloud-init status --long` under the top
level 'recoverable_errors' key as well as within the 'error' key nested
under the module-level keys: 'init-local', 'init', 'modules-config',
'modules-final'.

Implementation
==============

Cloud-init error codes
----------------------

.. code-block:: shell-session

    0 - success
    1 - unrecoverable error
    2 - recoverable error

If cloud-init exits with exit code 1, cloud-init experienced critical failure
and was unable to recover. In this case, something is likely seriously
wrong with the system, or cloud-init has experienced a serious bug.

If cloud-init exits with exit code 2, cloud-init was able to complete
gracefully, however something went wrong and the user should investigate.

Reported state
--------------
Cloud-init can report its internal state via the `status --format json`
subcommand under the `'extended_status'` key.

.. code-block:: shell-session

    $ cloud-init status --format json | jq .extended_status
    "degraded done"

See the list of all possible states:

> .. code-block:: shell-session

Author: context added

    "not running"
    "running"
    "done"
    "error"
    "degraded done"
    "degraded running"
    "disabled"

Exported errors: Aggregated errors
----------------------------------
When a recoverable error occurs, the internal cloud-init state
information is made visible under a top level aggregate key
'recoverable_errors' with errors sorted by error level:

.. code-block:: shell-session

    $ cloud-init status --format json | jq .recoverable_errors
    {
      "WARNING": [
        "Failed at merging in cloud config part from p-01: empty cloud config",
        "No template found in /etc/cloud/templates for template source.deb822",
        "No template found in /etc/cloud/templates for template sources.list",
        "No template found, not rendering /etc/apt/soures.list.d/ubuntu.source"
      ]
    }

Reported recoverable error messages are grouped by the level at which
they are logged. Complete list of levels:

> .. code-block:: shell-session

Author: context added (now in exported_errors.rst)

    WARNING
    DEPRECATED
    ERROR
    CRITICAL

Exported errors: Per-stage errors
---------------------------------
The keys 'errors' and 'recoverable_errors' are also exported for each
stage to allow attribution of recoverable and non-recoverable errors
to their source.

.. code-block:: shell-session

    $ cloud-init status --format json | jq .init.recoverable_errors
    {
      "WARNING": [
        "Failed at merging in cloud config part from p-001: empty cloud config"
      ]
    }

Note: Only cloud-init stages which have completed are listed in the
output of `cloud-init status --format json`.

The json representation of cloud-init stages (in run order) is:

> .. code-block:: shell-session

Author: agreed, thanks!

> .. _failure_states:

Reviewer 2: I'm not seeing any of the code blocks being rendered in this file. Nothing gets rendered at all in their place.

> Failure states: recoverable errors vs non-recoverable errors

Reviewer 2: Maybe just `Failure states`? This is kind of a large section heading

============================================================

> critical failure

Reviewer 2: ```suggestion
Critical failure
```

----------------
If cloud-init is unable to complete, error messages will now

> additionally be visible in output of `cloud-init status --format json`

Reviewer 2: Mentioned elseware, but I think we should prefer `--long`

within the 'error' key nested under the module-level keys: 'init-local',
'init', 'modules-config', 'modules-final'.

> recoverable failure

Reviewer 2: ```suggestion
Recoverable failure
```

-------------------
In the case that cloud-init is able to complete yet something goes awry,
the service will now return with exit code 2, and error messages will be
visible in the output of `cloud-init status --format` json under the top
level 'recoverable_errors' key as well as within the 'error' key nested
under the module-level keys: 'init-local', 'init', 'modules-config',
'modules-final'.

Implementation
==============

Cloud-init error codes
----------------------

>  0 - success

Reviewer 2: These 3 lines are rendering as a single line

 1 - unrecoverable error
 2 - recoverable error

If cloud-init exits with exit code 1, cloud-init experienced critical failure
and was unable to recover. In this case, something is likely seriously
wrong with the system, or cloud-init has experienced a serious bug.

If cloud-init exits with exit code 2, cloud-init was able to complete
gracefully, however something went wrong and the user should investigate.

Reported state
--------------
Cloud-init can report its internal state via the `status --format json`

> subcommand undert the `'extended_status'` key.

Reviewer 2: ```suggestion
subcommand under the `'extended_status'` key.
```

> $ cloud-init status --format json | jq .extended_status

Reviewer 2: This would be better rendered as a code block

"degraded done"

See the list of all possible states:

> .. code-block: shell-session

Reviewer 2: Having an inline list here but then appendices for the next two sections is a little odd, especially when this is the longest list. Personally, I prefer them all inline, but I'm fine either way as long as we are consistent.

.. _failure_states:

Failure states
==============

Cloud-init has multiple modes of failure. This page describes these
modes and how to gather information about failures.

Critical failure
----------------

Critical failures happens when cloud-init experiences a condition that it
cannot safely handle. When this happens, cloud-init may be unable to complete,
and the instance is likely to be in an unknown broken state.

Cloud-init experiences critical failure when:

* there is a major problem with the cloud image that is running cloud-init
* there is a severe bug in cloud-init

When this happens, error messages will be visible in output of
``cloud-init status --long`` within the ``'error'``.

The same errors will also be located under the key nested under the

> keys that store information related to each

Reviewer: ```suggestion
module-level keys that store information related to each
```

:ref:`stage of cloud-init<boot_stages>`: ``init-local``, ``init``,
``modules-config``, ``modules-final``.

Recoverable failure
-------------------

In the case that cloud-init is able to complete yet something went wrong,
cloud-init has experienced a "recoverable failure". When this happens,
the service will return with exit code 2, and error messages will be
visible in the output of ``cloud-init status --long`` under the top
level ``recoverable_errors`` and ``error`` keys.

To identify which stage an error came from, one can check under the
module-level keys: ``init-local``, ``init``, ``modules-config``,
``modules-final`` for the same error keys.

> See :ref:`here<exported_errors>` for to learn how to use cloud-init's
> exported errors.

Reviewer: ```suggestion
See :ref:`this more detailed explanation <exported_errors>` to learn how to use cloud-init's
exported errors.
```

Cloud-init error codes
----------------------

Cloud-init's ``status`` subcommand is useful for understanding which type of
error cloud-init experienced while running. The return code will be one of the
following:

.. code-block:: shell-session

    0 - success
    1 - unrecoverable error
    2 - recoverable error

If ``cloud-init status`` exits with exit code 1, cloud-init experienced
critical failure and was unable to recover. In this case, something is likely
seriously wrong with the system, or cloud-init has experienced a serious bug.
If you believe that you have experienced a serious bug, please file a
`bug report<reporting_bugs>`.

If cloud-init exits with exit code 2, cloud-init was able to complete
gracefully, however something went wrong and the user should investigate.

> See :ref:`here<reported_status>` for more information on cloud-init's status.

Reviewer: ```suggestion
See :ref:`this more detailed explanation <reported_status>` for more information on cloud-init's status.
```

Where to next?

> --------------
> See :ref:`here<how_to_debug>` for a detailed guide to debugging cloud-init.

Reviewer: ```suggestion
--------------

See :ref:`here<how_to_debug>` for a detailed guide to debugging cloud-init.
```

> .. _failure_states:

Author: Could be, but while related they are really different concepts. One is about the states of cloud-init, the other is related to extracting information if cloud-init is in one of those states. I'd rather keep them separate for now, but I'll add a link to the page since it isn't defined here - no need to re-define.
